123435 mhv portal incorrect link to profile

steveg-IAT opened the pull request:
## Are you removing, renaming or moving a folder in this PR?
- [x] No, I'm not changing any folders (skip to TeamSites and delete the rest of this section)
- [ ] Yes, I'm removing, renaming or moving a folder


Did you change site-wide styles, platform utilities or other infrastructure?
- [x] No
- [ ] Yes, and I used the [proxy-rewrite steps](https://github.com/department-of-veterans-affairs/vets-website/tree/main/src/applications/proxy-rewrite#that-sounds-normal-so-whats-the-proxy-all-about) to test the injected header scenario

## Summary
Email Confirmation Flow > user selects to add/edit on profile page
- Fixing the incorrect links with hash in the email alerts to target the email section heading (H2) of the profile/contact-info page
- refactored existing useEffect for focus and scrolling to page element to account for dynamic/async content load order.


## Related issue(s)

- https://github.com/department-of-veterans-affairs/va.gov-team/issues/123435
- https://github.com/department-of-veterans-affairs/va.gov-team/issues/122590

## Testing done

- unit test passed
- manually a11y tested in (VO) screenreader and with keyboard-only navigation
- Steps for local testing MyHealtheVet && Profile:

(mind your portnumber)

- http://localhost:3001/my-health/
- http://localhost:3001/profile/contact-information

**Force email alert**
- **selectors.js** ~ line 25:
/src/platform/mhv/components/MhvAlertConfirmEmail/selectors.js
  - add/uncomment out the `export const showAlert = state => true;` line
  - comment the original `export const showAlert = state => ...` block

```js
// export const showAlert = state => true;
export const showAlert = state =>
  !dismissedAlertViaCookie() &&
  !state.featureToggles.loading &&
  state.featureToggles.mhvEmailConfirmation &&
  !state.user.profile.loading &&
  state.user.profile.vaPatient &&
  (!selectContactEmailAddress(state) ||
    !selectContactEmailConfirmationDate(state) ||
    isBefore(
      new Date(selectContactEmailConfirmationDate(state)),
      new Date(DATE_THRESHOLD),
    ) ||
    !selectContactEmailUpdatedAt(state) ||
    isBefore(
      new Date(selectContactEmailUpdatedAt(state)),
      new Date(DATE_THRESHOLD),
    ));
```


## Screenshots

- screenshots in linked related tickets
- video of scroll function
https://github.com/user-attachments/assets/7cdc3ac8-9d76-4312-9ac9-909dbfb0207a
- video with VO screenreader on
https://github.com/user-attachments/assets/1e363cf1-6223-41ec-b41d-2849456dbfc0



## What areas of the site does it impact?

- https://staging.va.gov/my-va/
- https://staging.va.gov/my-health/
- https://staging.va.gov/profile/contact-information

## Acceptance criteria
- [ ] The links to profile/contact-information in the email add/confirmation alerts components properly target the correct page and  `#email-address` page id (H2 element) 
- [ ] Once the page/components load, the page scrolls to the defined page id `#email-address` and focus is set on this element.
  - Once scroll and focus occur, the screen reader (VO) will properly announce the current focus item H2 Email Address and allow the user to navigate to the next content or button on the page (*an additional alert in this case)
  - **Note** the profile page suppresses the yellow focus indicator on headings

### Quality Assurance & Testing

- [x] I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [x] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [x] Linting warnings have been addressed
- [ ] Documentation has been updated ([link to documentation](#) \*if necessary)
- [x] Screenshot of the developed feature is added (via related issue links)
- [x] [Accessibility testing](https://depo-platform-documentation.scrollhelp.site/collaboration-cycle/prepare-for-an-accessibility-staging-review) has been performed

### Error Handling

- [x] Browser console contains no warnings or errors.
- [ ] Events are being sent to the appropriate logging solution
- [ ] Feature/bug has a monitor built into Datadog or Grafana (if applicable)

### Authentication

- [ ] Did you login to a local build and verify all authenticated routes work as expected with a test user

steveg-IAT committed:
[123435] fixed incorrect page#target in href variables

steveg-IAT committed:
[123435]: refactored existing useEffect for focus and scrolling to account for dynamic content timing and order in loading.

steveg-IAT committed:
[123435]: fixed scroll issue introduced via eslint fix.

steveg-IAT committed:
[123435]: fixed focus after scroll onto #hashed element

steveg-IAT committed:
Refactor ContactInformation focus/scroll logic; add defensive checks;

steveg-IAT committed:
Refactor contact info deep-link focus logic for flexibility and test coverage

- Added a generic slug-based resolver for #edit-* deep links, supporting future buttons without code changes.
- Improved focus/scroll logic to handle any edit button or heading deep link, with fallback to the default focus target for unsupported hashes.
- Enabled and stabilized all deep-link Cypress E2E tests, including edit button and unsupported hash cases.

steveg-IAT committed:
Fix ContactInformation unit test for duplicate email matches (occurs with alrert present)

ckeif-iat commented on src/applications/personalization/profile/tests/components/contact-information/ContactInformation.unit.spec.jsx:
```diff
@@ -62,7 +62,7 @@ const testContactInfo = () => {
   // expect(view.getByText('214-718-2112', { exact: false })).to.exist;
 
-  expect(view.getByText(/alongusername/)).to.exist;
+  expect(view.getAllByText(/alongusername/)).to.have.length.above(0);
```
[question] Why did this change? I don't see us rendering anything new in the DOM in `ContactInformation.jsx` so just curious.

steveg-IAT replied:
adding comment...
`// check for multiple alongusername in email address when alert is rendered and contact email is also rendered (confirm + edit)`

steveg-IAT replied:
Re-adding, I must have gotten munged/removed/overlooked during the merge conflict nastiness.

ckeif-iat reviewed (dismissed):
Nice, LGTM 🔥 ! 

A few minor suggestions and a question, but nothing blocking.

carlosfelix2 commented on src/applications/personalization/profile/components/contact-information/ContactInformation.jsx:
```diff
@@ -24,6 +24,35 @@ const getScrollTarget = hash => {
+
+// Attempt to derive a matching va-button for any #edit-* hash by comparing normalized labels
+// This avoids brittle hard-coded label maps and automatically supports new edit buttons.
+const getEditButtonFromHash = hash => {
```
I like the idea of having unit tests for this specific thing.

carlosfelix2 commented on src/applications/personalization/profile/components/contact-information/ContactInformation.jsx:
```diff
@@ -65,50 +95,130 @@ const ContactInformation = () => {
     () => {
-      // Set the focus on the page's focus target _unless_ one of the following
-      // is true:
+      // Origininal:
```
There is no need for documenting the replaced code.

carlosfelix2 commented on src/applications/personalization/profile/components/contact-information/ContactInformation.jsx:
```diff
@@ -65,50 +95,130 @@ const ContactInformation = () => {
-      if (lastLocation?.pathname.match(new RegExp(pathRegExp))) {
-        return;
+      //   a11y reasonse
+      // New: refactored to improve reliability of focus and scroll behavior for hash-based navigation
```
Don't explain here your improvements. Do that in the PR description.

carlosfelix2 commented on src/applications/personalization/profile/components/contact-information/ContactInformation.jsx:
```diff
@@ -65,50 +95,130 @@ const ContactInformation = () => {
+      //   return undefined;
+      // }
+
+      const cameFromProfileRoot = Boolean(
```
There is a large amount of code there, but no real tests to make sure this works. What is your code coverage here?

carlosfelix2 reviewed (changes requested):
You code coverage is too low for `profile/components/contact-information/ContactInformation.jsx‎`. Please add more tests.

> 1.88% Statements 49/117, 14.92% Branches 10/67, 58.33% Functions 14/24, 47.52% Lines 48/101

steveg-IAT committed:
123435: fixed incorrect page#target in href variables introduced in merge

steveg-IAT committed:
123435: extract deep-link hash utilities from ContactInformation

- Move toSlug, getEditButtonFromHash, getScrollTarget from ContactInformation.jsx to util/hashUtils.js
- Add comprehensive JSDoc comments explaining why utilities exist and how they work
- Update ContactInformation.jsx imports to use extracted utilities
- Remove inline utility implementations and commented legacy code
- Improve maintainability by centralizing hash-based deep-linking logic

steveg-IAT committed:
123435: fixing PR review items "missed" after reset/merge issues

steveg-IAT commented on src/applications/personalization/profile/tests/e2e/contact-information/deep-links.cypress.spec.js:
```diff
@@ -70,9 +70,14 @@ function checkAllDeepLinks(mobile = false) {
     // focus should be managed correctly
-    // console.log('expectedTarget', expectedTarget);
-    getTargetElement(expectedTarget).should('have.focus');
+    cy.log('profile path: ', PROFILE_PATHS.CONTACT_INFORMATION);
```
Interesting, not at all what my line:75 has... will make sure file gets updated.

steveg-IAT committed:
123435: MHV: Portal: post PR review additions and cleanup

steveg-IAT committed:
updated/removed disable eslint comment (again)
